micro-rdk/src/esp32/entry.rs

```diff
@@ -337,55 +369,105 @@ async fn serve_async_with_external_network<S>(
+
+            let mut duration = None;
+            loop {
+                if let Some(duration) = duration {
+                    Timer::after(duration).await;
+                } else {
+                    // TODO: Maybe some back-off up to a limit
+                    let _ = duration.insert(Duration::from_secs(3));
                 }
-                // Assume connected to internet so any error should be forwarded to provisioning
-                if let Err(e) = validate_robot_credentials(
+
+                log::info!("Attempting to validate stored robot credentials");
+                match validate_robot_credentials(
                     exec.clone(),
                     &storage.get_robot_credentials().unwrap(),
+                    &mut client_connector,
                 )
                 .await
                 {
-                    if let Err(e) = storage.reset_robot_credentials() {
-                        log::error!("couldn't erase credentials {:?}", e);
+                    Ok(app_client) => {
+                        log::info!("Robot credentials validated OK");
+                        break 'provisioned app_client;
+                    }
+                    Err(e) => {
+                        if let Some(app_client_error) = e.downcast_ref::<AppClientError>() {
+                            if matches!(app_client_error, AppClientError::AppGrpcClientError(GrpcClientError::GrpcError{ code, .. }) if *code == 7 || *code == 16)
+                            {
+                                // The validate call failed with an explicit rejection (PERMISSION_DENIED/UNAUTHENTICATED)
+                                // of the credentials. Reset the cached credentials and any robot configuration, and
+                                // move on to provisioning.
+                                log::warn!("Robot credential validation failed permanently with error {:?}; initiating provisioning", e);
+
+                                if let Err(e) = storage.reset_robot_credentials() {
+                                    log::error!("Couldn't erase robot credentials {:?}", e);
+                                }
+
+                                if let Err(e) = storage.reset_robot_configuration() {
+                                    log::error!("couldn't erase robot configuration {:?}", e);
+                                }
+
+                                // Record the last error so that we can serve it once we reach provisioning.
+                                let _ = last_error.insert(e);
+                                break;
+                            }
+                        }
+
+                        // For all other errors, we assume we could not communicate with app due
+                        // to network issues, and just restart the inner loop until we are able
+                        // to communicate with app.
+                        log::info!("Unable to validate robot credentials {:?}; will retry", e);
```

Comment:
this means that as long as we don't have network connectivity at start the robot won't run?

Reply:
No, if there are both cached credentials and configuration present the robot will be created at line 397 (new). This loop will then spin attempting to communicate with app to validate. While the loop is spinning the robot will be collecting data. If/when app becomes reachable, if the robot validates then we are provisioned and we move on to `serve_web_inner`. Otherwise, if validation fails due to an explicit rejection of the credentials, the robot gets dropped and we move on to provisioning. Or, it is considered some sort of transient networking error and the loop keeps looping. Or, at least that is the intention!